registry/remote/auth/example_test.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+// Package remote_test includes all the testable examples for remote repository type
+
+package auth_test
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"reflect"
+	"strings"
+	"testing"
+
+	"oras.land/oras-go/v2/registry/remote/auth"
+)
+
+const (
+	username    = "test_user"
+	password    = "test_password"
+	accessToken = "test/access/token"
+)
+
+var host string
+var simpleDir string
+var basicAuthDir string
+var accessTokenDir string
+var clientConfigDir string
+var scopes = []string{
+	"repository:dst:pull,push",
+	"repository:src:pull",
+}
+
+func TestMain(m *testing.M) {
+	// create an authorization server
+	as := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != http.MethodGet {
+			w.WriteHeader(http.StatusUnauthorized)
+			panic("unexecuted attempt of authorization service")
+		}
+		header := "Basic " + base64.StdEncoding.EncodeToString([]byte(username+":"+password))
+		if auth := r.Header.Get("Authorization"); auth != header {
+			w.WriteHeader(http.StatusUnauthorized)
+		}
+		if got := r.URL.Query().Get("service"); got != host {
+			w.WriteHeader(http.StatusUnauthorized)
+		}
+		if got := r.URL.Query()["scope"]; !reflect.DeepEqual(got, scopes) {
+			w.WriteHeader(http.StatusUnauthorized)
+		}
+		if _, err := fmt.Fprintf(w, `{"access_token":%q}`, accessToken); err != nil {
+			panic(err)
+		}
+	}))
+	defer as.Close()
+
+	// create a test server
+	ts := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		path := r.URL.Path
+		if r.Method != http.MethodGet {
+			w.WriteHeader(http.StatusNotFound)
+			panic("unexpected access")
+		}
+		switch path {
+		case "/basicAuth":
+			wantedAuthHeader := "Basic " + base64.StdEncoding.EncodeToString([]byte(username+":"+password))
+			authHeader := r.Header.Get("Authorization")
+			if authHeader != wantedAuthHeader {
+				w.Header().Set("Www-Authenticate", `Basic realm="Test Server"`)
+				w.WriteHeader(http.StatusUnauthorized)
+			}
+		case "/accessToken":
+			wantedAuthHeader := "Bearer " + accessToken
+			if auth := r.Header.Get("Authorization"); auth != wantedAuthHeader {
+				challenge := fmt.Sprintf("Bearer realm=%q,service=%q,scope=%q", as.URL, host, strings.Join(scopes, " "))
+				w.Header().Set("Www-Authenticate", challenge)
```

Comment:
The www-authenticat header won't be used if access token is used, see https://github.com/oras-project/oras-go/blob/90fab5e123f99bd5da9ba261321cab02575e7fa6/registry/remote/auth/client.go#L263

Reply:
For this and the authorization server comment, please refer to the spec  in https://msazure.visualstudio.com/AzureContainerRegistry/_git/specs?path=/Registry/Access-Control/runtime.md&anchor=token-auth&_a=preview

Reply:
I wrote this part according to this unit test: https://github.com/oras-project/oras-go/blob/90fab5e123f99bd5da9ba261321cab02575e7fa6/registry/remote/auth/client_test.go#L254